### PR TITLE
Add activity notification library

### LINKED ISSUE #243: Notification for users.
**Is your feature request related to a problem? Please describe.**
with commontator subscriptions, mentions, upvote, downvote, and groups it's important to make a system to notify the user about these inside the website.  

**Describe the solution you'd like**
 A notification system similar to GitHub would just be a perfect solution but we can start with just a basic notification option in navbar which can show us our notifications from notification table in which we can push whenever any group invite is sent or any comment mention is made.

**Describe alternatives you've considered**
I looked into [notifications](https://github.com/rails-engine/notifications) gem which seems a decent enough option I am looking for a better alternative though.
 
**Are you working on this? (Yes/No)**
If you guys think this is the right time to implement this feature I can start working on this.

### PR DESCRIPTION
**Pull request by junron**
Fixes #243 

#### Describe the changes you have made in this pr -

- [x] Added [activity notification](https://github.com/simukappu/activity_notification) library
- [x] Added notification for starring project
- [x] Added tests
- [x] Link to notifications page
- [x] Web push
- [x] Icon for web push

#### Demonstration video (Web push notifications)
output
[MP4 on google drive](https://drive.google.com/open?id=162zJd4RKtAcwKtOQe0SinsGtNZW5TaV2)

Users enable browser notifications by visiting the Notifications page and clicking enable. The browser will ask them for permission and if granted, generate encryption and authentication keys, as well as a cloud endpoint for sending notifications to the device, which are sent to the server and stored in the database. This process needs to be repeated for every device that the user wants to receive push notifications on. The user can then click the test button to send a test notification to itself.

The server also needs to identify itself to the browser's cloud push notification endpoint. It does this by generating a public/private key pair. You can do this by running `Webpush.generate_key` in the rails console:
```
$ rails console
Running via Spring preloader in process 8655
Loading development environment (Rails 6.0.1)
[1] pry(main)> Webpush.generate_key
=> #<Webpush::VapidKey:2ab51f32dbe8 :public_key=BAAS5N5HuL4ElT1hBU1ZdGfRrXq1Ja90gOR7IjwZXfSU4ZfogrIMVtnerHgsI2TLoidFPSH32_H89thI1l9Rdn4= :private_key=XPwmYJgtRybLJDWXNKIDVOdgXQzJzl0Uj29sw_52Bwo=>
```
These keys need to be exposed to the application by setting the following environment variables:
- `VAPID_PUBLIC_KEY`
- `VAPID_PRIVATE_KEY`

One final step is to add the icon for the push notifications. Currently, I did not specify an icon, so it defaults to the browser logo (in chrome at least). Please advise as to which image to use for the icon.

Ping @tachyons 

Replacement for #706

### CODE REVIEW

File: app/assets/javascripts/serviceworker.js.erb [hunk 1]
```diff
@@ -0,0 +1,71 @@
...
+  console.log('[Serviceworker]', "Installing!", event);
+  event.waitUntil(
+    caches.open(CACHE_NAME).then(function prefill(cache) {
+      return cache.addAll([
+
+        // make sure serviceworker.js is not required by application.js
+        // if you want to reference application.js from here
+        '<%#= asset_path "application.js" %>',
+
+        '<%= asset_path "application.css"
```
Comment (tachyons) [inline app/assets/javascripts/serviceworker.js.erb]: Are you sure this is not application.scss ?

File: app/assets/javascripts/serviceworker.js.erb [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+          return response;
+        }
+      })
+    })
+  );
+}
+
+self.addEventListener('install', onInstall);
+self.addEventListener('activate', onActivate);
+ self.addEventListener('fetch', onFetch);
```
Comment (tachyons) [inline app/assets/javascripts/serviceworker.js.erb]: remove extra space here

File: app/assets/javascripts/webpush.js.erb [hunk 3]
```diff
@@ -0,0 +1,47 @@
+<% ENV["VAPID_PUBLIC_KEY"] = ENV["VAPID_PUBLIC_KEY"] ||
+  "BEdmrrlB8FEHyvVKkzLaUPTQpIIPjqYB9A8fitbhDz-ZBgpbY3wYL1VY9M0Kz3vPVbRRQONpTXGYIioKullYZf4="
```
Comment (tachyons) [inline app/assets/javascripts/webpush.js.erb]: Can we move this to config/environments/development.rb ?

File: app/views/activity_notification/mailer/default/batch_default.html.erb [hunk 4]
```diff
@@ -0,0 +1,87 @@
...
+  <%= yield :head
+ %>
+</head>
+<body>
+<p>Dear <%= @target.printable_target_name
+ %></p>
+<% @notifications.each
+
+
+ do |notification| %>
```
Comment (tachyons) [inline app/views/activity_notification/mailer/default/batch_default.html.erb]: Don't make this multi line,

<% @notifications.each do |notification| %> looks much cleaner

File: config/initializers/activity_notification.rb [hunk 5]
```diff
@@ -0,0 +1,82 @@
...
+  config.subscription_enabled = false
+
+  # Configure default subscription value to use when the subscription record does not configured.
+  # Note that you can configure them for each method calling as default argument.
+  # Set false when you want to unsubscribe to any notifications as default.
+  config.subscribe_as_default = true
+
+  # Configure the e-mail address which will be shown in ActivityNotification::Mailer,
+  # note that it will be overwritten if you use your own mailer class with default "from" parameter.
+  config.mailer_sender = 'please-change-me-at-config-initializers-activity_notification@example.com'
```
Comment (tachyons) [inline config/initializers/activity_notification.rb]: Need to set this value

File: config/environments/development.rb [hunk 6]
```diff
@@ -1,4 +1,5 @@
 Rails.application.configure do
+  config.hosts << "circuitverse.chatbox2.ml"
```
Comment (tachyons) [inline config/environments/development.rb]: intentional ?

File: spec/controllers/push_subscription_controller_spec.rb [hunk 7]
```diff
@@ -0,0 +1,26 @@
...
+    @user = FactoryBot.create(:user)
+    sign_in @user
+  end
+
+  describe "#create" do
+    it "creates a push subscription" do
+      post "/push/subscription/new", params: { push_subscription: {
+          # Fake endpoint, for tests only
+          endpoint: "https://fcm.googleapis.com/fcm/send/",
+          auth: "GVWfxR22Bomw-uqi32NeBQ",
```
Comment (tachyons) [inline spec/controllers/push_subscription_controller_spec.rb]: hardcoded ?

File: app/views/activity_notification/mailer/default/batch_default.html.erb (hunk 4 above)
Comment (junron) [reply app/views/activity_notification/mailer/default/batch_default.html.erb]: Ok, I'll be undoing the erblint changes soon

File: spec/models/push_subscription_test.rb [hunk 8]
```diff
@@ -0,0 +1,9 @@
+# frozen_string_literal: true
+
+require "test_helper"
+
+class PushSubscriptionTest < ActiveSupport::TestCase
```
Comment (tachyons) [inline spec/models/push_subscription_test.rb]: This need to be in rspec

File: app/assets/javascripts/webpush.js.erb (hunk 3 above)
Comment (junron) [reply app/assets/javascripts/webpush.js.erb]: It is required for the tests I think

Comment (tachyons) [reply app/assets/javascripts/webpush.js.erb]: Then you can add it in config/environments/test.rb as well

File: spec/models/push_subscription_test.rb (hunk 8 above)
Comment (junron) [reply spec/models/push_subscription_test.rb]: Done

File: spec/controllers/push_subscription_controller_spec.rb (hunk 7 above)
Comment (junron) [reply spec/controllers/push_subscription_controller_spec.rb]: I've changed them all to "dummy"

**Comment by junron** [comment]
Everything resolved! @tachyons

File: app/models/star.rb [hunk 9]
```diff
@@ -1,4 +1,23 @@
+# frozen_string_literal: true
+
 class Star < ApplicationRecord
+  require "custom_optional_targets/web_push"
```
Comment (tachyons) [inline app/models/star.rb]: move require statement to outside class declaration

File: app/views/activity_notification/notifications/default/_default.html.erb [hunk 10]
```diff
@@ -0,0 +1,182 @@
...
+  <%# if notification.unopened? %>
+  <%#= link_to "Open and move (GET)", move_notification_path_for(notification, parameters.slice(:routing_scope, :devise_default_routes).merge(open: true)) %>
+  <%#= link_to "Open and move (POST)", open_notification_path_for(notification, parameters.slice(:routing_scope, :devise_default_routes).merge(move: true)), method: :post %>
+  <%#= link_to "Open", open_notification_path_for(notification, parameters.slice(:routing_scope, :devise_default_routes).merge(index_options: @index_options))), method: :post %>
+  <%#= link_to "Open (Ajax)", open_notification_path_for(notification, parameters.slice(:routing_scope, :devise_default_routes).merge(reload: false)), method: :post, remote: true %>
+  <%# end %>
+  <%#= link_to "Destroy", notification_path_for(notification, parameters.slice(:routing_scope, :devise_default_routes).merge(index_options: @index_options)), method: :delete %>
+  <%#= link_to "Destroy (Ajax)", notification_path_for(notification, parameters.slice(:routing_scope, :devise_default_routes).merge(reload: false)), method: :delete, remote: true %>
```
Comment (tachyons) [inline app/views/activity_notification/notifications/default/_default.html.erb]: Remove commented code

File: app/views/activity_notification/notifications/default/_default_without_grouping.html.erb [hunk 11]
```diff
@@ -0,0 +1,160 @@
...
+  <%# if notification.unopened? %>
+    <%#= link_to "Open and move (GET)", move_notification_path_for(notification, parameters.slice(:routing_scope, :devise_default_routes).merge(open: true)) %>
+    <%#= link_to "Open and move (POST)", open_notification_path_for(notification, parameters.slice(:routing_scope, :devise_default_routes).merge(move: true)), method: :post %>
+    <%#= link_to "Open", open_notification_path_for(notification, parameters.slice(:routing_scope, :devise_default_routes).merge(index_options: @index_options)), method: :post %>
+    <%#= link_to "Open (Ajax)", open_notification_path_for(notification, parameters.slice(with_group_members:, :routing_scope, :devise_default_routes).merge(reload: false)), method: :post, remote: true %>
+  <%# end %>
+  <%#= link_to "Destroy", notification_path_for(notification, index_options: parameters.slice(:routing_scope, :devise_default_routes).merge(index_options: @index_options)), method: :delete %>
+  <%#= link_to "Destroy (Ajax)", notification_path_for(notification, parameters.slice(with_group_members:, :routing_scope, :devise_default_routes).merge(reload: false)), method: :delete, remote: true %>
```
Comment (tachyons) [inline app/views/activity_notification/notifications/default/_default_without_grouping.html.erb]: same here

File: lib/custom_optional_targets/web_push.rb [hunk 12]
```diff
@@ -0,0 +1,16 @@
+# frozen_string_literal: true
+
+module CustomOptionalTarget
```
Comment (tachyons) [inline lib/custom_optional_targets/web_push.rb]: Can you move this to app/lib folder ?

File: spec/controllers/notifications_controller_spec.rb [hunk 13]
```diff
@@ -0,0 +1,24 @@
+# frozen_string_literal: true
+
+require "rails_helper"
+
+describe Users::NotificationsController, type: :request do
+  it "should get index page" do
+    get root_path
+    expect(response.status).to eq(200)
+  end
+  describe "Authentication" do
```
Comment (tachyons) [inline spec/controllers/notifications_controller_spec.rb]: A space after it block

File: spec/models/push_subscription_test.rb [hunk 14]
```diff
@@ -0,0 +1,9 @@
+# frozen_string_literal: true
+
+require "rails_helper"
+
+class PushSubscriptionTest < ActiveSupport::TestCase
```
Comment (tachyons) [inline spec/models/push_subscription_test.rb]: use describe syntaxt

**Comment by tachyons** [comment]
Great work,  Can someone from @CircuitVerse/gci-mentors help reviewing the ui and the functionality?

**Comment by junron** [comment]
@tachyons 
All issues pointed out resolved! 

Mentors can test out my PR at https://demo.circuitverse.chatbox2.ml/. The database has been seeded with `rails db:seed` so you can login with the default credentials. 

The icon issue has still now been resolved.

**Comment by junron** [comment]
@tachyons Could you advise on a suitable icon to use for the push notifications? It should be square and be at least 140px by 140px

**Comment by junron** [comment]
I am using this image:
image

**Comment by shubhankarsharma00** [comment]
@junron this is amazing work :+1:

**Comment by nitin10s** [comment]
Looks great😍

**Comment by junron** [comment]
Thank you all for the kind comments, I really learnt a lot from this task!

**Pull request merged**